added configurable parentElem and preventButton

bumblehead opened the pull request:
added configurable parentElem and preventButton

https://github.com/borismus/webvr-boilerplate/issues/68
https://github.com/borismus/webvr-boilerplate/issues/66

if defined, `parentElem` is used in place of window to obtain width and height. `preventButton` prevents button construction.

``` javascript
    var manager = new WebVRManager(renderer, effect, {
      parentElem : videoElem.parentNode,
      preventButton : true
    });
```

bumblehead committed:
added configurable parent element

borismus commented on src/webvr-manager.js:
```diff
@@ -46,12 +46,14 @@ function WebVRManager(renderer, effect, params) {
   this.effect = effect;
   this.distorter = new CardboardDistorter(renderer);
-  this.button = new ButtonManager();
+  this.button = !preventButton && new ButtonManager();
```
I don't like this conditionally created ButtonManager. Let's keep it always created.

borismus commented on src/webvr-manager.js:
```diff
@@ -104,10 +106,12 @@ function WebVRManager(renderer, effect, params) {
-    this.button.on('vr', this.onVRClick_.bind(this));
-    this.button.on('back', this.onBackClick_.bind(this));
-    this.button.on('settings', this.onSettingsClick_.bind(this));
+    if (this.button) {
```
Mainly because these conditionals just complicate things.

bumblehead replied:
If you don't want to support options for changing or disabling the buttons, I agree with you.

borismus commented on src/webvr-manager.js:
```diff
@@ -46,12 +46,14 @@ function WebVRManager(renderer, effect, params) {
+  var preventButton = this.params.preventButton || false;
+  
   // Save the THREE.js renderer and effect for later.
+  this.parentElem = params.parentElem;
```
how about `parentEl` or better yet, `parentElement`

bumblehead replied:
ok

borismus commented on src/webvr-manager.js:
```diff
@@ -329,16 +336,30 @@ WebVRManager.prototype.anyModeToNormal_ = function() {
+WebVRManager.prototype.getParentElem_ = function () {
+  var elem = this.parentElem;
+  
+  return elem && elem instanceof Element ? elem : window;
```
Why not do this type-checking in the constructor?

bumblehead replied:
I'm not totally satisfied with this myself. If you want to continue using a constructor-defined reference to the element, I agree with you.

borismus commented on src/webvr-manager.js:
```diff
@@ -329,16 +336,30 @@ WebVRManager.prototype.anyModeToNormal_ = function() {
 WebVRManager.prototype.resize_ = function() {
-  this.effect.setSize(window.innerWidth, window.innerHeight);
+  var elem = this.getParentElem_(),
+      width = typeof elem.innerWidth === 'number' ? elem.innerWidth : elem.offsetWidth,
```
Why use innerWidth with a fallback to offsetWidth? In what cases is innerWidth unavailable, but outerWidth available?

bumblehead replied:
document.documentElement is defined with offsetWidth but not innerWidth. window is defined with innerWidth but not offsetWidth.

bumblehead commented:
@borismus thank you for reviewing and considering my PR with your agreeable response.

Changes I made are motivated by subjectiveness. For example, window is defined with innerWidth and document elements are defined with offsetWidth which in my experience is more consistently handled by browsers than outerWidth. 

In my full preference an element (not window) would always be used as the parent because this would be simpler than supporting both window or element. The parent element would always be accessed from the DOM with an id value to ensure the parent element is still present in the document at each time reference is needed.

For those who would build their own system of video player controls, it would be best if the boilerplate ui were separated and more easily bypassed.

borismus commented:
Please address my inline comments, and rebase against latest changes.

bumblehead commented:
I rebased my branch named parentElem-configure, https://github.com/bumblehead/webvr-boilerplate/tree/parentElem-configure

I removed the button related changes, but added a new change which I hope you will agree with. this.parentEl is params.parentEl _or_ document.body (window is no longer used).

bumblehead committed:
remove call to undefined getParentElem_

bumblehead committed:
update to match latest webvr-boilerplate master source

borismus commented:
Outdated

abuisine commented:
Hi there :)
I was pretty much interested in this PR (manager sized to a parent div for instance)
can I give an hand on this ?
what is the plan ?

borismus commented:
Please rebase and make small focused changes. The original PR here had a lot of seemingly unrelated changes.

edsilv commented:
Hi,

I'm also interested in this issue and agree with what @bumblehead says about always using a dom element as the parent and separating out the UI.

As mentioned by @pindiespace in https://github.com/borismus/webvr-polyfill/issues/23 it sounds like a significant overhaul is needed. To second @abuisine, what is the plan? Perhaps if we can split it up into sub-tasks we can each take one on?

borismus commented:
This is in large part solved with 1.0 WebVR polyfill.

edsilv commented:
Cool, do you mean on this branch?

https://github.com/borismus/webvr-polyfill/tree/webvr-1.0

Should I be using the webvr-1.0 branch of webvr-boilerplate as well?

Presumably I can specify a parent element somehow? 

This is the main roadblock I've run into for my 'virtex' viewer. I will be demonstrating this at the MoMA in a couple of weeks and would love to be able to show people webvr-boilerplate-enabled VR.

borismus commented:
The webvr branch was merged to master for a while, so no need for the other
branch.
You could use the polyfill directly rather than using the boilerplate. Let
me know if you need help.

On Mon, Apr 25, 2016 at 8:15 AM Edward Silverton notifications@github.com
wrote:

> Cool, do you mean on this branch?
> 
> https://github.com/borismus/webvr-polyfill/tree/webvr-1.0
> 
> Should I be using the webvr-1.0 branch of webvr-boilerplate as well?
> 
> Presumably I can specify a parent element somehow?
> 
> This is the main roadblock I've run into for my 'virtex' viewer. I will be
> demonstrating this at the MoMA in a couple of weeks and would love to be
> able to show people webvr-boilerplate-enabled VR.
> 
> —
> You are receiving this because you modified the open/close state.
> 
> Reply to this email directly or view it on GitHub
> https://github.com/borismus/webvr-boilerplate/pull/77#issuecomment-214387637

edsilv commented:
I've commented all uses of webvr-boilerplate, just using the polyfill now.

Based on your example I've tried to recreate the setup in Viewport.ts. If I uncomment this line:

https://github.com/edsilv/virtex/blob/webvr/src/Viewport.ts#L406

the camera location seems to be shifted to 0, 0, 0.

Any idea what I'm doing wrong?

edsilv commented:
to clarify, the camera should be offset like this:

http://universalviewer.io/examples/?manifest=http%3A%2F%2Ffiles.universalviewer.io%2Fmanifests%2Ffoundobjects%2Fthekiss.json

uncommenting 

```
this._vrControls.update();
```

puts the camera into the center of the object

edsilv commented:
aha

https://github.com/mrdoob/three.js/blob/master/examples/js/controls/VRControls.js#L81

edsilv commented:
That seems to be working now:

http://edsilv.github.io/virtex/test/

Ignore the broken loading bar, it works, it just doesn't like gzipped content coming from github.

This has external controls and can toggle between vr/normal mode. One hack I had to put in is this:

https://github.com/edsilv/virtex/blob/gh-pages/test/js/VREffect.js#L72

(forced `isPresenting = true`)

I need to figure out how to debug this on my android phone (I have a Google Cardboard). Any tips gratefully received :-)

borismus commented:
I recommend chrome:inspect. Why did you need the hack? Are you calling
requestPresent?

On Wed, Apr 27, 2016 at 1:57 AM Edward Silverton notifications@github.com
wrote:

> That seems to be working now:
> 
> http://edsilv.github.io/virtex/test/
> 
> Ignore the broken loading bar, it works, it just doesn't like gzipped
> content coming from github.
> 
> This has external controls and can toggle between vr/normal mode. One hack
> I had to put in is this:
> 
> https://github.com/edsilv/virtex/blob/gh-pages/test/js/VREffect.js#L72
> 
> (forced isPresenting = true)
> 
> I need to figure out how to debug this on my android phone (I have a
> Google Cardboard). Any tips gratefully received :-)
> 
> —
> You are receiving this because you modified the open/close state.
> 
> Reply to this email directly or view it on GitHub
> https://github.com/borismus/webvr-boilerplate/pull/77#issuecomment-215017304
